# Projeto da aula #: Análise Exploratória de Dados de Doenças Cardíacas

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ry4n-felinto/impytech_2026/blob/main/aula_05/projeto_biomedicina.ipynb)

Desenvolvedores: 

* Gabriel Ferreira Silva (al.gabriel.silva@impatech.edu.br)
* Pedro Rodrigues Alberti (al.pedro.alberti@impatech.edu.br)

---

## Sobre este projeto

Você vai trabalhar com a base **Heart Failure Prediction** (Kaggle), que reúne
918 linhas compostas por registros clínicos de pacientes com 11 variáveis mais 
o desfecho (presença ou não de doença cardíaca). O objetivo  é praticar o fluxo 
completo de uma análise exploratória: baixar os dados, inspecionar, limpar, resumir 
e visualizar.

### Objetivos de aprendizagem
- Baixar, carregar e inspecionar um `DataFrame` com pandas
- Identificar problemas simples nos dados (nulos, valores impossíveis)
- Construir alguns gráficos com Altair e interpretá-los

---

## 0. Setup — importar as bibliotecas

Estaremos utilizando as bibliotecas `pandas` e `altair` para o seguinte notebook. Além disso, vamos aumentar o número máximo de colunas processados pelo Altair para não termos problemas:

In [3]:
import pandas as pd
import altair as alt

# Deixa o Altair usar bases um pouco maiores sem reclamar (opcional).
alt.data_transformers.enable("default", max_rows=5000)

print("pandas:", pd.__version__)
print("altair:", alt.__version__)

pandas: 3.0.3
altair: 6.2.2


---

## 1. Baixar os dados

Baixe os dados no site do Kaggle em https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction e direcione o caminho do arquivo.


No exemplo abaixo, temos que o caminho indicado é `data/heart.csv`, mas caso seja de seu interesse, pode ser qualquer diretório válido e que faça sentido:

In [5]:
CAMINHO_CSV = "data/heart.csv"

---

## 2. Carregar e conhecer os dados

Vamos dar uma olhada nos dados que temos nessa base de dados, analisando as colunas e os valores que temos. Para isso, utilizamos a função `read_csv` da biblioteca Pandas:

In [6]:
df = pd.read_csv(CAMINHO_CSV)
df.head()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0


Quantas linhas e colunas temos? Que tipo de dado existe em cada coluna? Podemos fazer isso utilizando o `df.shape`:

In [7]:
# Quantas linhas e colunas? Que tipo tem cada coluna?
print("Formato (linhas, colunas):", df.shape)
df.info()

Formato (linhas, colunas): (918, 12)
<class 'pandas.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    str    
 2   ChestPainType   918 non-null    str    
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    str    
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    str    
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    str    
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), str(5)
memory usage: 86.2 KB


Vamos ver as estatísticas descritivas das colunas numéricas utilizando o método `describe()` no nosso dataframe:

In [ ]:
df.describe()

,Age,RestingBP,Cholesterol,FastingBS,MaxHR,Oldpeak,HeartDisease
count,918.000000,918.000000,918.000000,918.000000,918.000000,918.000000,918.000000
mean,53.510893,132.396514,198.799564,0.233115,136.809368,0.887364,0.553377
std,9.432617,18.514154,109.384145,0.423046,25.460334,1.066570,0.497414
min,28.000000,0.000000,0.000000,0.000000,60.000000,-2.600000,0.000000
25%,47.000000,120.000000,173.250000,0.000000,120.000000,0.000000,0.000000
50%,54.000000,130.000000,223.000000,0.000000,138.000000,0.600000,1.000000
75%,60.000000,140.000000,267.000000,0.000000,156.000000,1.500000,1.000000
max,77.000000,200.000000,603.000000,1.000000,202.000000,6.200000,1.000000


### Dicionário de variáveis

Temos as seguintes explicações de variáveis retiradas do site do Kaggle.

| Coluna | Significado | Tipo |
|---|---|---|
| `Age` | Idade (anos) | numérica |
| `Sex` | Sexo (M/F) | categórica |
| `ChestPainType` | Tipo de dor no peito (ATA, NAP, ASY, TA) | categórica |
| `RestingBP` | Pressão arterial em repouso (mmHg) | numérica |
| `Cholesterol` | Colesterol sérico (mg/dl) | numérica |
| `FastingBS` | Glicemia de jejum > 120 mg/dl (1=sim, 0=não) | binária |
| `RestingECG` | Eletrocardiograma em repouso (Normal, ST, LVH) | categórica |
| `MaxHR` | Frequência cardíaca máxima atingida | numérica |
| `ExerciseAngina` | Angina induzida por exercício (Y/N) | categórica |
| `Oldpeak` | Depressão do segmento ST | numérica |
| `ST_Slope` | Inclinação do segmento ST (Up, Flat, Down) | categórica |
| `HeartDisease` | **Desfecho:** 1 = doença cardíaca, 0 = normal | binária |


---

## 3. Limpeza e preparação

Antes de analisarmos, é importante você verificar dois pontos:
1. Existem valores nulos ou impossíveis ( valores que 'não fazem sentido' )?
2. _Extra_: O que poderíamos fazer para "substituir" esses valores nulos? Faz sentido trocarmos esses valores por outra coisa? Podemos ignorar os dados com esses valores nulos? 


Sobre o segundo ponto. Veremos uma abordagem estatística para este problema nas próximas 2 aulas.


In [9]:
# 1. Valores nulos por coluna
df.isna().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [10]:
# 2. Quantos valores iguais a zero em colunas onde isso é impossível?
print("Colesterol = 0:", (df["Cholesterol"] == 0).sum())
print("Pressão (RestingBP) = 0:", (df["RestingBP"] == 0).sum())

Colesterol = 0: 172
Pressão (RestingBP) = 0: 1


---

## 4. Análise exploratória

A partir de agora, a sequência do trabalho está em suas mãos. Se questione perguntas simples e diretas e tente responde-las fazendo análises numéricas dos dados. De exemplo, temos as seguintes perguntas que podem ser exploradas:

* Existe um gênero com maiores tendências cardíacas?
* A idade tem uma influência maior de ter a doenças cardíacas?
* Pessoas com glicemia em jejum elevada (FastingBS = 1) têm maior incidência de problemas cardíacos?
* Qual a taxa de doença cardíaca para cada tipo de dor (ChestPainType)?
* Existe alguma variável que não parece fazer alguma influência para predizer o problema analizado?

Para a primeira pergunta, temos um exemplo de análise simples abaixo, junto com uma breve explicação dos achados:

In [11]:
# Quantos pacientes têm e não têm doença cardíaca?
df["HeartDisease"].value_counts()

HeartDisease
1    508
0    410
Name: count, dtype: int64

In [12]:
# Proporção de doença cardíaca por sexo
pd.crosstab(df["Sex"], df["HeartDisease"], normalize="index")

HeartDisease,0,1
Sex,,
F,0.740933,0.259067
M,0.368276,0.631724


Pela tabela acima, podemos ver que cerca de 63% dos homens possuem doenças cardíacas, comparado aos 25% das mulheres com problemas de coração, o que indica que pessoas do gênero masculino possuem mais chances de terem problemas cardíacos.

In [ ]:
# Idade x problemas cardíacos

In [ ]:
# Glicemia x problemas cardíacos

In [ ]:
# ChestPainType X problemas cardíacos

In [ ]:
# Outros X problemas cardíacos

---

## 5. Visualizações com Altair

Visualizações tem por objetivo **facilitar** o entendimento dos dados, se fazendo muito importante em contextos com maiores valores de icógnitas. Imagine por exemplo, que queremos ver a distribuição de frequência cardíaca máxima pela idade, agora temos diversas combinações de valor das variáveis. E se eu quisesse ver a relação desses 2 **preditores** com a presença de doença cardíaca? 

### 5.3 Idade x frequência cardíaca máxima, colorido pelo desfecho

Vamos fazer um scatter-plot utilizando as variáveis `Age` (Idade), `MaxHR` (Frequência cardíaca máxima) e `HeartDisease` (Problemas cardíacos). Utilizando as codifiações dos eixos X e Y para idade e frequência cardíaca máxima, respectivamente.

In [13]:
alt.Chart(df).mark_circle(size=60, opacity=0.5).encode(
    x=alt.X("Age:Q", title="Idade"),
    y=alt.Y("MaxHR:Q", title="Freq. cardíaca máxima"),
    color=alt.Color("HeartDisease:N", title="Doença (1=sim)"),
    tooltip=["Age", "MaxHR", "Sex", "ChestPainType"],
).interactive().properties(title="Idade x MaxHR", width=500)

alt.Chart(...)

É possível ver que não existe uma divisão clara de idade e frequência cardíaca máxima, visto que os dois grupos ainda estão entrelaçados entre si.

Faça algumas visualizações com as colunas do dataframe junto com a idade, utilizando a codificação de cor `alt.Color`. Não se esqueça que essa análise só faz sentido para variáveis numéricas.

In [ ]:
# visualizações: colesterol x idade.

---

## 6. Conclusões

Escreva de 3 a 5 frases resumindo o que você encontrou. Alguns pontos de partida:
- Qual variável mais parece se relacionar com o desfecho?
- Alguma descoberta te surpreendeu?
- Que cuidado com os dados (ex.: colesterol = 0) você teve de tomar?

_(Escreva aqui)_


---
### Finalização
1. Salve o notebook (`Ctrl+S`).
2. Coloque o arquivo em seu repositório próprio.
3. Em caso de dúvidas ou até feedback, entre em contato com os professores e lembre-se de encaminhar junto o link do repositório **com visibilidade pública**.
